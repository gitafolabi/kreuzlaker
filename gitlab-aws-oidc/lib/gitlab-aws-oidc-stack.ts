import {Stack, StackProps} from "aws-cdk-lib";
import {Construct} from "constructs";
import * as iam from "aws-cdk-lib/aws-iam";
import * as cdk from 'aws-cdk-lib';

export interface GitlabAWSOIDCStackProps extends StackProps {
    projectName: string;
    allowedRepoToPush: string;
    allowedBranchesToPush: string[];
    gitlabHost: string;
    thumbprints: string[];
}

export class GitlabAWSOIDCStack extends Stack {
    public gitlabPipelineRole: iam.Role;

    constructor(scope: Construct, id: string, props: GitlabAWSOIDCStackProps) {
        super(scope, id, props);
        const {projectName, allowedRepoToPush, allowedBranchesToPush, gitlabHost, thumbprints} = props;

        const gitlabOIDCProvider = new iam.OpenIdConnectProvider(
            this,
            "GitlabPipeline",
            {
                // See https://docs.gitlab.com/ee/ci/cloud_services/aws/
                url: `https://${gitlabHost}`,
                clientIds: [`https://${gitlabHost}`],
                thumbprints: thumbprints,
            }
        );

        const allowedBranchPatternToPush: string[] = []

        // See https://docs.gitlab.com/ee/ci/cloud_services/aws/ for how the pattern needs to look like
        for (let branch of allowedBranchesToPush) {
            allowedBranchPatternToPush.push(`project_path:${allowedRepoToPush}:ref_type:branch:ref:${branch}`)
        }

        const stringEqualsCondition: {
            [key: string]: string[]
        } = {}
        stringEqualsCondition[`${gitlabHost}:sub`] = allowedBranchPatternToPush

        this.gitlabPipelineRole = new iam.Role(this, "GitlabPipelineRole", {
            // Trust policy
            assumedBy: new iam.WebIdentityPrincipal(
                gitlabOIDCProvider.openIdConnectProviderArn,
                {
                    StringEquals: stringEqualsCondition,
                }
            ),
            roleName: `gitlab-oidc-${projectName}`,
            description: `Role to assume from gitlab pipeline of ${projectName}`,
        });

        // Configure the role to allow whatever the CICD pipeline should be able to do.
        // In this case, allow this role to assume the cdk roles, to run cdk diff/deploy
        this.gitlabPipelineRole.addToPolicy(iam.PolicyStatement.fromJson({
                    "Effect": "Allow",
                    "Action": ["sts:AssumeRole"],
                    "Resource": [
                        "arn:aws:iam::*:role/cdk-*-lookup-role-*",
                        "arn:aws:iam::*:role/cdk-*-image-publishing-role-*",
                        "arn:aws:iam::*:role/cdk-*-file-publishing-role-*",
                        "arn:aws:iam::*:role/cdk-*-deploy-role-*",
                        // this has full administrator access!
                        // "arn:aws:iam::*:role/cdk-*-cfn-exec-role-*",
                    ]
                }
            )
        )

        // for pushing a docker image to the corresponding ECR repo
        // https://docs.aws.amazon.com/AmazonECR/latest/userguide/image-push.html
        this.gitlabPipelineRole.addToPolicy(iam.PolicyStatement.fromJson({
                    "Effect": "Allow",
                    "Action": [
                        "ecr:CompleteLayerUpload",
                        "ecr:UploadLayerPart",
                        "ecr:InitiateLayerUpload",
                        "ecr:BatchCheckLayerAvailability",
                        "ecr:PutImage"
                    ],
                    "Resource": `arn:aws:ecr:${this.region}:${this.account}:repository/dbt-run`
                }
            )
        );
        this.gitlabPipelineRole.addToPolicy(iam.PolicyStatement.fromJson({
                    "Effect": "Allow",
                    "Action": "ecr:GetAuthorizationToken",
                    "Resource": "*"
                }
            )
        )

        // output the role arn for easier copy and paste into the gitlab pipeline
        new cdk.CfnOutput(this, 'PipelineRoleArn', {
            value: this.gitlabPipelineRole.roleArn,
            description: `The arn of the role which can be assumed from a gitlab pipeline on https://${gitlabHost} within the following repos+branches: ${allowedBranchPatternToPush}`,
            exportName: 'PipelineRoleArn',
        });
    }
}
