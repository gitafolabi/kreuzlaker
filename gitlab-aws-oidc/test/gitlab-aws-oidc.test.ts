import * as cdk from "aws-cdk-lib";
import {Template, Match} from "aws-cdk-lib/assertions";
import {GitlabAWSOIDCStack} from "../lib/gitlab-aws-oidc-stack";

describe("test cdk stack", () => {
    const app = new cdk.App();
    const projectName = "whatever";
    const allowedRepoToPush = "department-name/team-name/repo-name"
    const allowedBranchesToPush: string[] = [
        "main",
        "develop",
    ];

    const stackWithTwoBranchPattern = new GitlabAWSOIDCStack(app, "MyTestStackWithTwoBranchPattern", {
        projectName,
        allowedRepoToPush,
        allowedBranchesToPush: [
            "main",
            "develop",
        ],
        gitlabHost: "gitlab.example.com",
    });

    const stackWithSingleBranchPattern = new GitlabAWSOIDCStack(
        app,
        "MyTestStackWithOnlyOneBranchPattern",
        {
            projectName,
            allowedRepoToPush,
            allowedBranchesToPush: ["main"],
            gitlabHost: "gitlab.example.com",
        }
    );

    test("OIDC Provider created", () => {
        const template = Template.fromStack(stackWithTwoBranchPattern);

        template.hasResourceProperties("Custom::AWSCDKOpenIdConnectProvider", {
            Url: "https://gitlab.example.com",
            ClientIDList: ["https://gitlab.example.com"],
        });
    });

    test("IAM role with multiple branch patterns exists", () => {
        const template = Template.fromStack(stackWithTwoBranchPattern);

        template.hasResourceProperties("AWS::IAM::Role", {
            AssumeRolePolicyDocument: {
                Statement: [
                    {
                        Action: "sts:AssumeRoleWithWebIdentity",
                        Condition: {
                            StringEquals: {
                                "gitlab.example.com:sub": [
                                    "project_path:department-name/team-name/repo-name:ref_type:branch:ref:main",
                                    "project_path:department-name/team-name/repo-name:ref_type:branch:ref:develop",
                                ],
                            },
                        },
                        Effect: "Allow",
                    },
                ],
            },
        });
    });

    test("IAM role with single branch pattern exists", () => {
        const template = Template.fromStack(stackWithSingleBranchPattern);

        template.hasResourceProperties("AWS::IAM::Role", {
            AssumeRolePolicyDocument: {
                Statement: [
                    {
                        Action: "sts:AssumeRoleWithWebIdentity",
                        Condition: {
                            StringEquals: {
                                "gitlab.example.com:sub": [
                                    "project_path:department-name/team-name/repo-name:ref_type:branch:ref:main",
                                ],
                            },
                        },
                        Effect: "Allow",
                    },
                ],
            },
        });
    });

    test("IAM Policy for OIDC role allows required actions", () => {
        const template = Template.fromStack(stackWithSingleBranchPattern);
        const pipelineRole = stackWithSingleBranchPattern.gitlabPipelineRole
        const resolvedGitlabPipelineRoleName = stackWithSingleBranchPattern.resolve(pipelineRole.roleName)

        template.hasResourceProperties("AWS::IAM::Policy", {
            "Roles": Match.arrayWith(resolvedGitlabPipelineRoleName),
            // change the next to what your gitlab deployment pipeline needs
            "PolicyDocument": {
                "Statement": [
                    {
                        "Action": "sts:AssumeRole",
                        "Effect": "Allow",
                        "Resource": [
                            "arn:aws:iam::*:role/cdk-*-lookup-role-*",
                            "arn:aws:iam::*:role/cdk-*-image-publishing-role-*",
                            "arn:aws:iam::*:role/cdk-*-file-publishing-role-*",
                            "arn:aws:iam::*:role/cdk-*-deploy-role-*",
                        ]
                    }
                ],
            },
        });
    });

    test('Snapshot of whole stack stays stable', () => {
        // This needs adjustments for almost every change to the stack, so it's mostly good to test refactorings
        // and check if the actual diff makes sense
        // to update the snapshot: run `npm test -- -u`
        const template = Template.fromStack(stackWithSingleBranchPattern);
        expect(template.toJSON()).toMatchSnapshot();
    });
});