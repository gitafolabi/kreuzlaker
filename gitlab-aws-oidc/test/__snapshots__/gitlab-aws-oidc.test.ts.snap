// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test cdk stack Snapshot of whole stack stays stable 1`] = `
Object {
  "Outputs": Object {
    "PipelineRoleArn": Object {
      "Description": "The arn of the role which can be assumed from a gitlab pipeline on https://gitlab.example.com within the following repos+branches: project_path:department-name/team-name/repo-name:ref_type:branch:ref:main",
      "Export": Object {
        "Name": "PipelineRoleArn",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "GitlabPipelineRole8B3B9E26",
          "Arn",
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0": Object {
      "DependsOn": Array [
        "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "494e82802f093e633e580188c7937aef9d6ee0ad4a788d58c647e6bea26d2915.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "iam:CreateOpenIDConnectProvider",
                    "iam:DeleteOpenIDConnectProvider",
                    "iam:UpdateOpenIDConnectProviderThumbprint",
                    "iam:AddClientIDToOpenIDConnectProvider",
                    "iam:RemoveClientIDFromOpenIDConnectProvider",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GitlabPipelineF8D7C83D": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ClientIDList": Array [
          "https://gitlab.example.com",
        ],
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0",
            "Arn",
          ],
        },
        "Url": "https://gitlab.example.com",
      },
      "Type": "Custom::AWSCDKOpenIdConnectProvider",
      "UpdateReplacePolicy": "Delete",
    },
    "GitlabPipelineRole8B3B9E26": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "StringEquals": Object {
                  "gitlab.example.com:sub": Array [
                    "project_path:department-name/team-name/repo-name:ref_type:branch:ref:main",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": Object {
                  "Ref": "GitlabPipelineF8D7C83D",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role to assume from gitlab pipeline of whatever",
        "RoleName": "gitlab-oidc-whatever",
      },
      "Type": "AWS::IAM::Role",
    },
    "GitlabPipelineRoleDefaultPolicyF178FD85": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Array [
                "arn:aws:iam::*:role/cdk-*-lookup-role-*",
                "arn:aws:iam::*:role/cdk-*-image-publishing-role-*",
                "arn:aws:iam::*:role/cdk-*-file-publishing-role-*",
                "arn:aws:iam::*:role/cdk-*-deploy-role-*",
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GitlabPipelineRoleDefaultPolicyF178FD85",
        "Roles": Array [
          Object {
            "Ref": "GitlabPipelineRole8B3B9E26",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
